nitorch/tools/affine_reg.py

```diff
@@ -0,0 +1,961 @@
+        grid = _affine2grid(grid0, M)
+        # Resample to fixed grid
+        dat_new = grid_pull(dat[m][None, None, ...], grid[None, ...],
+            bound='dft', extrapolate=True, interpolation=1)[0, 0, ...]
+        if cost_fun == 'njtv':
+            # Add to NJTV cost for moving image
+            mtv += dat_new
+            njtv =- dat_new.sqrt()
+
+    # Compute the cost function
+    res = None
+    if cost_fun in costs_hist:  # Histogram based costs
+        # Compute joint histogram
+        H = _hist_2d(dat_fix, dat_new)
+        res = H
+
+        # Get probabilities
+        pxy = H / H.sum()
+        px = pxy.sum(dim=0, keepdim=True)
+        py = pxy.sum(dim=1, keepdim=True)
+
+        # Compute cost
+        if cost_fun == 'mi':
+            # Mutual information
+            mi = torch.sum(pxy * torch.log2(pxy / py.mm(px)))
+            c = -mi
+        elif cost_fun == 'ecc':
+            # Entropy Correlation Coefficient
+            mi = torch.sum(pxy * torch.log2(pxy / py.mm(px)))
+            ecc = -2*mi/(torch.sum(px*px.log2()) + torch.sum(py*py.log2()))
+            c = -ecc
+        elif cost_fun == 'nmi':
+            # Normalised Mutual Information
+            nmi = (torch.sum(px*px.log2()) + torch.sum(py*py.log2()))/torch.sum(pxy*pxy.log2())
+            c = -nmi
+        elif cost_fun == 'ncc':
+            # Normalised Cross Correlation
+            i = torch.arange(1, pxy.shape[0] + 1, device=device, dtype=dtype)
+            j = torch.arange(1, pxy.shape[1] + 1, device=device, dtype=dtype)
+            m1 = torch.sum(py*i[..., None])
+            m2 = torch.sum(px*j[None, ...])
+            sig1 = torch.sqrt(torch.sum(py[..., 0]*(i - m1)**2))
+            sig2 = torch.sqrt(torch.sum(px[0, ...]*(j - m2)**2))
+            i, j = torch.meshgrid(i - m1, j - m2)
+            ncc = torch.sum(torch.sum(pxy*i*j))/(sig1*sig2)
+            c = -ncc
+    elif cost_fun == 'njtv':  # Total variation based cost
+        njtv += torch.sqrt(N)*mtv.sqrt()
+        res = njtv
+        c = torch.sum(njtv)
+
+    if was_numpy:
+        # Back to numpy array
+        c = c.cpu().numpy()
+
+    if return_res:
+        return c, res
+    else:
+        return c
+
+
+def _data_loader(imgs, opt):
+    """Load image data, affine matrices, image dimensions.
+
+    Parameters
+    ----------
+    imgs : list
+        See run_affine_reg.
+    opt : dict
+        See run_affine_reg.
+
+    Returns
+    ----------
+    dat : [N,] tensor_like
+        List of image data.
+    mats : [N,] tensor_like
+        List of affine matrices.
+    dims : [N,] tensor_like
+        List of image dimensions.
+
+    """
+    if opt['cost_fun'] in costs_hist:
+        mx_int = 255
+    elif opt['cost_fun'] == 'njtv':
+        mx_int = None
+
+    dat = []
+    mats = []
+    dims = []
+    for img in imgs:
+        # Load data
+        dat_p, M_p, _, _, _ = load_3d(img,
+            samp=0, do_mask=False, truncate=True, mx_out=mx_int,
```

Comment:
OK, that's where you rescale your data to [0, 255]. Maybe add a comment to state that it's important, as the rest of the code expects this.

Reply:
Done